lxd/placement/filtering.go

```diff
@@ -0,0 +1,243 @@
+	"github.com/canonical/lxd/shared/api"
+)
+
+// GetInstancesInPlacementGroup returns a map of instance ID to node ID for all instances that reference the given
+// placement group either directly or indirectly via a profile.
+func GetInstancesInPlacementGroup(ctx context.Context, tx *sql.Tx, project string, placementGroupName string) (map[int]int, error) {
+	// Query notes:
+	// 1. Union of profile config and instance config to perform single query. Since we plan to run this on the leader,
+	//    this may not be necessary.
+	// 2. The apply order must be a selected column, as we need to perform the ORDER BY to expand config correctly, and
+	//    the ORDER BY can only be used after the UNION statement.
+	// 3. The apply_order of 1000000 for instances is used to ensure that instance config is applied last.
+	q := `
+SELECT
+	instances.id,
+	instances.node_id,
+	profiles_config.value,
+	instances_profiles.apply_order AS apply_order
+FROM instances 
+	JOIN projects ON instances.project_id = projects.id
+	JOIN instances_profiles ON instances.id = instances_profiles.instance_id 
+	JOIN profiles ON instances_profiles.profile_id = profiles.id 
+	JOIN profiles_config ON profiles.id = profiles_config.profile_id
+WHERE projects.name = ? AND profiles_config.key = 'placement.group'
+UNION
+SELECT 
+	instances.id,
+	instances.node_id,
+	instances_config.value,
+	1000000 AS apply_order
+FROM instances
+	JOIN projects ON instances.project_id = projects.id
+	JOIN instances_config ON instances.id = instances_config.instance_id 
+WHERE projects.name = ? AND instances_config.key = 'placement.group'
+	ORDER BY instances.id, apply_order`
+	args := []any{project, project}
+
+	// Keep a map of pointers so that each value is mutable.
+	instIDToPlacementGroup := make(map[int]string)
+	instIDToNodeID := make(map[int]int)
+	err := query.Scan(ctx, tx, q, func(scan func(dest ...any) error) error {
+		var instID int
+		var nodeID int
+		var placementGroup string
+		var applyOrder int
+		err := scan(&instID, &nodeID, &placementGroup, &applyOrder)
+		if err != nil {
+			return err
+		}
+
+		instIDToNodeID[instID] = nodeID
+		instIDToPlacementGroup[instID] = placementGroup
+		return nil
+	}, args...)
+	if err != nil {
+		return nil, err
+	}
+
+	result := make(map[int]int, len(instIDToPlacementGroup))
+	for id, group := range instIDToPlacementGroup {
+		if group == placementGroupName {
+			result[id] = instIDToNodeID[id]
+		}
+	}
+
+	return result, nil
+}
+
+// Filter filters the candidates using the placement group. If the excludeInstanceIDs argument is passed, those instances
+// are ignored (this is used to check if a running instance satisfies its placement group policy, or during evacuation).
+func Filter(ctx context.Context, tx *sql.Tx, candidates []db.NodeInfo, excludeInstanceIDs []int, placementGroup cluster.PlacementGroup) ([]db.NodeInfo, error) {
+	// Get the placement group config.
+	apiPlacementGroup, err := placementGroup.ToAPI(ctx, tx)
```

Comment:
rather than calling `ToAPI` here, I'd prefer to see the `apiPlacementGroup` passed in as an argument, this way it makes clear what the function needs to operate, and the caller is left to convert `placementGroup` to `apiPlacementGroup`, this way it can avoid needing to do this multiple times if calling `Filter` multiples times.